src/messaging/ReliableMessageManager.cpp

```diff
@@ -404,57 +449,29 @@ CHIP_ERROR ReliableMessageManager::SendFromRetransTable(RetransTableEntry * entr
     CHIP_ERROR err              = CHIP_NO_ERROR;
     ReliableMessageContext * rc = entry->rc;
 
-    // To trigger a call to OnSendError, set the number of transmissions so
-    // that the next call to ExecuteActions will abort this entry,
-    // restart the timer immediately, and ExitNow.
-
-    CHIP_FAULT_INJECT(FaultInjection::kFault_RMPSendError, entry->sendCount = static_cast<uint8_t>(rc->mConfig.mMaxRetrans + 1);
-                      entry->nextRetransTimeTick = 0; StartTimer(); ExitNow());
-
     if (rc)
     {
-        // Locally store the start and length;
-        uint8_t * p  = entry->msgBuf->Start();
-        uint16_t len = entry->msgBuf->DataLength();
-
-        // Send the message through
-        uint16_t msgSendFlags = entry->msgSendFlags;
-        SetFlag(msgSendFlags, MessageFlagValues::kMessageFlag_RetainBuffer);
-        err = SendMessage(rc, entry->msgBuf, msgSendFlags);
+        err = mSessionMgr->SendMessage(entry->payloadHeader, entry->peerNodeId, std::move(entry->retainedBuf), &entry->retainedBuf);
 
-        // Reset the msgBuf start pointer and data length after sending
-        entry->msgBuf->SetStart(p);
-        entry->msgBuf->SetDataLength(len);
+        if (err == CHIP_NO_ERROR)
+        {
+            // Update the counters
+            entry->sendCount++;
+        }
+        else
+        {
+            // Remove from table
+            ChipLogError(ExchangeManager, "Crit-err %ld when sending CHIP MsgId:%08" PRIX32 ", send tries: %d", long(err),
+                         entry->retainedBuf.GetMsgId(), entry->sendCount);
 
-        // Update the counters
-        entry->sendCount++;
+            ClearRetransTable(*entry);
+        }
     }
     else
     {
         ChipLogError(ExchangeManager, "Table entry invalid");
     }
 
-    VerifyOrExit(err != CHIP_NO_ERROR, err = CHIP_NO_ERROR);
-
-    // Any error generated during initial sending is evaluated for criticality which would
-    // qualify it to be reportable back to the caller. If it is non-critical then
-    // err is set to CHIP_NO_ERROR.
```

Comment:
Now we relay on the return value of this function to know if the entry->retainedBuf is updated since its initial value is nil

mSessionMgr->SendMessage(entry->payloadHeader, entry->peerNodeId, std::move(entry->retainedBuf), &entry->retainedBuf);

Only the return value is CHIP_NO_ERROR, we can guarantee the entry->retainedBuf is properly updated.

Reply:
Ah, that is a good point.  Maybe file a followup issue to figure out whether we need to deal with non-critical errors here? If we do, we'll need to think about a slightly different API somehow...

Reply:
https://github.com/project-chip/connectedhomeip/issues/4218